tests/unit/test_driver_yaml.py

```diff
@@ -9,114 +8,110 @@
-
-        class D4(BaseCrafter):
-            pass
-
-        d4 = D4()
-        assert len(d4._drivers) == 4
-
-        class D5(BaseCrafter):
-            pass
-
-        d5 = D5()
-        assert len(d5._drivers) == 4
-
-
-if __name__ == '__main__':
-    unittest.main()
+def test_load_yaml1(tmpdir):
+    with open(os.path.join(cur_dir, 'yaml/test-driver.yml'), encoding='utf8') as fp:
+        a = yaml.load(fp)
+
+    assert isinstance(a[0], KVSearchDriver)
+    assert isinstance(a[1], ControlReqDriver)
+    assert isinstance(a[2], BaseDriver)
+
+    with open(os.path.join(tmpdir, 'test_driver.yml'), 'w', encoding='utf8') as fp:
+        yaml.dump(a[0], fp)
+
+    with open(os.path.join(tmpdir, 'test_driver.yml'), encoding='utf8') as fp:
+        b = yaml.load(fp)
+
+    assert isinstance(b, KVSearchDriver)
+    assert b._executor_name == a[0]._executor_name
+
+
+def test_load_cust_with_driver():
+    a = BaseExecutor.load_config('mwu-encoder/mwu_encoder_driver.yml')
+    assert a._drivers['ControlRequest'][0].__class__.__name__ == 'MyAwesomeDriver'
+    p = set_pod_parser().parse_args(['--uses', os.path.join(cur_dir, 'mwu-encoder/mwu_encoder_driver.yml')])
+    with Pod(p):
+        # will print a cust task_name from the driver when terminate
+        pass
+
+
+def test_pod_new_api_from_kwargs():
+    a = BaseExecutor.load_config('mwu-encoder/mwu_encoder_driver.yml')
+    assert a._drivers['ControlRequest'][0].__class__.__name__ == 'MyAwesomeDriver'
+
+    with Pod(uses=os.path.join(cur_dir, 'mwu-encoder/mwu_encoder_driver.yml')):
+        # will print a cust task_name from the driver when terminate
+        pass
+
+
+def test_load_yaml2(tmpdir):
+    os.environ['JINA_TEST_EXEC_WITH_DRIVER'] = str(tmpdir)
+    a = BaseExecutor.load_config(os.path.join(cur_dir, 'yaml/test-exec-with-driver.yml'))
+    assert len(a._drivers) == 2
+    # should be able to auto fill in ControlRequest
+    assert 'ControlRequest' in a._drivers
+    a.save_config()
+    p = a.config_abspath
+    b = BaseExecutor.load_config(p)
+    assert a._drivers == b._drivers
+    a.touch()
+    a.save()
+    c = BaseExecutor.load(a.save_abspath)
+    assert a._drivers == c._drivers
+    del os.environ['JINA_TEST_EXEC_WITH_DRIVER']
+
+
+def test_resource_executor():
+    a = BaseExecutor.load_config(resource_filename('jina', '/'.join(('resources', 'executors._route.yml'))))
+    assert a.name == 'route'
+    assert len(a._drivers) == 4
+    a = BaseExecutor.load_config(resource_filename('jina', '/'.join(('resources', 'executors._pass.yml'))))
+    assert a.name == 'forward'
+    assert len(a._drivers) == 4
+    a = BaseExecutor.load_config(resource_filename('jina', '/'.join(('resources', 'executors._merge.yml'))))
+    assert a.name == 'merge'
+    assert len(a._drivers) == 4
+    a = BaseExecutor.load_config(resource_filename('jina', '/'.join(('resources', 'executors._clear.yml'))))
+    assert a.name == 'clear'
+    assert len(a._drivers) == 4
```

Comment:
we can parameterise this one

Reply:
Would u mind doing a PR to this branch to see how u would proceed?

Reply:
```python
import pytest

@pytest.mark.parametrize(
    'yaml, name, expected',
    [
        ('executors._route.yml', 'route', 4),
        ('executors._pass.yml',  'forward', 4),
        # the rest ..
    ]
)
def test_resource_executor(yaml, name, expected):
    a = BaseExecutor.load_config(resource_filename('jina', '/'.join(('resources', yaml))))
    assert a.name == name
    assert len(a._drivers) == expected
```

Reply:
(I guess this is easier)

Reply:
Done! Thanks!